app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt

```diff
@@ -32,4 +37,59 @@ sealed class Diagnostic {
                 append(" Expected symbols: ${expectedSymbols.joinToString()}.")
         }.toString()
     }
+
+    data class ConstantWithoutValue(val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "A constant must have a value"
+    }
+
+    data class UninitializedGlobalVariable(val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "A global variable must be initialized"
+    }
+
+    data class ImmutableAssignment(val assignment: Statement.Assignment, val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a " + if (variable.kind == Variable.Kind.CONSTANT) "constant" else "value"
+    }
+
+    data class ParameterAssignment(val assignment: Statement.Assignment, val parameter: Function.Parameter) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a parameter"
+    }
+
+    data class FunctionAssignment(val assignment: Statement.Assignment, val function: Function) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a function"
+    }
+
+    data class FunctionAsValue(val expression: Expression.Variable, val function: Function) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot use a function as a value"
+    }
+
+    data class VariableCall(val call: Expression.FunctionCall, val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot call a variable"
+    }
+
+    data class ParameterCall(val call: Expression.FunctionCall, val parameter: Function.Parameter) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot call a parameter"
+    }
+
+    data class ConditionalTypesMismatch(val conditional: Expression.Conditional, val typeWhenTrue: Type, val typeWhenFalse: Type) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "The results of a conditional operator cannot have distinct types '$typeWhenTrue' and '$typeWhenFalse'"
+    }
+
+    data class NonConstantExpression(val expression: Expression) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Expected a constant expression"
+    }
+
+    data class InvalidType(val expression: Expression, val type: Type, val expectedType: Type) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Expected type '$expectedType' instead of '$type'"
+    }
```

Comment:
To remain consistent with the others, could you put those diagnostics in some sealed class, e. g. `TypeCheckingError`?

Reply:
Done.

Reply:
Done.